api/src/services/application-exporter.service.ts

```diff
@@ -0,0 +1,764 @@
+              },
+              where: {
+                multiselectQuestionId: null,
+              },
+              orderBy: {
+                ordinal: OrderByEnum.DESC,
+              },
+            }
+          : undefined,
+      },
+      where: {
+        listingId: queryParams.id,
+        deletedAt: null,
+        markedAsDuplicate: forLottery ? false : undefined,
+      },
+    });
+    console.log('basic app data:', process.memoryUsage());
+    // get all multiselect questions for a listing to build csv headers
+    const multiSelectQuestions =
+      await this.multiselectQuestionService.findByListingId(queryParams.id);
+
+    // get maxHouseholdMembers associated to the selected applications
+    let maxHouseholdMembers = 0;
+    applications.forEach((app) => {
+      if (app.householdMember?.length > maxHouseholdMembers) {
+        maxHouseholdMembers = app.householdMember.length;
+      }
+    });
+
+    const columns = getExportHeaders(
+      maxHouseholdMembers,
+      multiSelectQuestions,
+      queryParams.timeZone,
+      queryParams.includeDemographics,
+      forLottery,
+    );
+
+    if (forLottery) {
+      applications = applications.filter(
+        (elem) => !!elem.applicationLotteryPositions?.length,
+      );
+
+      applications = applications.sort(
+        (a, b) =>
+          a.applicationLotteryPositions[0].ordinal -
+          b.applicationLotteryPositions[0].ordinal,
+      );
+    }
+    const mappedApps = mapTo(Application, applications);
+    console.log('before raw sheet gen:', process.memoryUsage());
+    await this.generateSpreadsheetData(
+      workbook,
+      mappedApps,
+      columns,
+      queryParams,
+      forLottery,
+    );
+
+    if (forLottery) {
+      const preferences = multiSelectQuestions.filter(
+        (question) =>
+          question.applicationSection ===
+          MultiselectQuestionsApplicationSectionEnum.preferences,
+      );
+      for (const preference of preferences) {
+        console.log(`before ${preference.text} gen:`, process.memoryUsage());
+        await this.generateSpreadsheetData(
+          workbook,
+          mappedApps,
+          columns,
+          queryParams,
+          forLottery,
+          {
+            id: preference.id,
+            name: preference.text,
+          },
+        );
+      }
+    }
+  }
+
+  /**
+   *
+   * @param workbook the spreadsheet we'll be adding data too
+   * @param applications the full list of partial applications
+   * @param csvHeaders the headers and renderers of the export
+   * @param queryParams the incoming param args
+   * @param forLottery whether we are getting the lottery results or not
+   * @param preference if present, then builds the preference specific spreadsheet page
+   * @returns void but writes the output to a file
+   */
+  async generateSpreadsheetData(
+    workbook: Excel.Workbook,
+    applications: Application[],
+    csvHeaders: CsvHeader[],
```

Comment:
nit: since this is no longer a csv, does the name csvHeader still make sense?

Reply:
ah, those are actually the csvHeaders and we take those in and convert them to the spreadsheet headers in the helper function `buildExportColumns()` so I think thats still appropriate but open to changing it if you feel its still too confusing!